### PR TITLE
Check for read before init for a single @Initializer method

### LINKED ISSUE #61: Check for read before init for a single `@Initializer` method
Follow-on to #57.  When a class has a single `@Initializer` method, we know it runs after all the initializer blocks / constructors.  So, we should be able to check that in the `@Initializer` no `@NonNull` fields are read unless they were already initialized locally or in some constructor / initializer block.  With multiple `@Initializer` methods we don't know what order they run in, so less checking is possible.

### PR DESCRIPTION
**Pull request by msridhar**
Fixes #61

### CODE REVIEW

**Comment by msridhar** [comment]
This PR is still work in progress.  Will be pushing patches to fix more bugs.  Will post a comment when I think I'm done 😄

File: gradle.properties [hunk 1]
```diff
@@ -13,7 +13,7 @@
 # org.gradle.parallel=true
 
 GROUP=com.uber.nullaway
-VERSION_NAME=0.1.9-SNAPSHOT
+VERSION_NAME=0.1.9-LOCAL
```
Comment (lazaroclapp) [inline gradle.properties]: Should we be committing this? Seems like an unrelated change at best :)

File: nullaway/src/main/java/com/uber/nullaway/NullAway.java [hunk 2]
```diff
@@ -519,16 +525,41 @@ private Description checkForReadBeforeInit(ExpressionTree tree, VisitorState sta
...
+    return checkPossibleUninitFieldRead(tree, state, symbol, path, enclosingBlockPath);
   }
 
-  private boolean relevantInitializerMethodOrBlock(Tree methodLambdaOrBlock, VisitorState state) {
+  private Symbol.ClassSymbol enclosingClassSymbol(TreePath enclosingBlockPath) {
+    Tree leaf = enclosingBlockPath.getLeaf();
+    if (leaf instanceof BlockTree) {
+      // parent must be a ClassTree
+      Tree parent = enclosingBlockPath.getParentPath().getLeaf();
+      return ASTHelpers.getSymbol((ClassTree) parent);
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/NullAway.java]: Just out of curiosity, this is an optimization over the else branch, right? `ASTHelpers.enclosingClass(ASTHelpers.getSymbol(leaf))` will always get the enclosing class and is only more expensive?

File: nullaway/src/test/resources/com/uber/nullaway/testdata/ReadBeforeInitNegativeCases.java [hunk 3]
```diff
@@ -159,4 +164,114 @@ private void initF() {
...
+
+  static class CastToNonNullTest {
+
+    Object castF;
+    Object castG;
+
+    @Initializer
+    void init(@Nullable Object o) {
+      if (o != null) {
+        castF = castToNonNull(castF);
```
Comment (lazaroclapp) [inline nullaway/src/test/resources/com/uber/nullaway/testdata/ReadBeforeInitNegativeCases.java]: Cool.

File: nullaway/src/main/java/com/uber/nullaway/NullAway.java [hunk 4]
```diff
@@ -694,14 +713,60 @@ private boolean fieldInitializedByPreviousInitializer(
...
+      BinaryTree binaryTree = (BinaryTree) parent;
+      Tree.Kind kind = binaryTree.getKind();
+      if (kind.equals(Tree.Kind.EQUAL_TO) || kind.equals(Tree.Kind.NOT_EQUAL_TO)) {
+        ExpressionTree left = binaryTree.getLeftOperand();
+        ExpressionTree right = binaryTree.getRightOperand();
+        return (left.equals(leaf) && right.getKind().equals(Tree.Kind.NULL_LITERAL))
+            || (right.equals(leaf) && left.getKind().equals(Tree.Kind.NULL_LITERAL));
+      }
+    } else if (parent instanceof MethodInvocationTree) {
+      // ok if it's invoking castToNonNull and the read is the argument
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/NullAway.java]: How about passing it to other methods that take `@Nullable`? Do we care to allow that case?

Comment (lazaroclapp) [reply nullaway/src/main/java/com/uber/nullaway/NullAway.java]: (I don't mean in this PR, just a possible extension)

File: nullaway/src/main/java/com/uber/nullaway/NullAway.java (hunk 2 above)
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/NullAway.java]: No the issue is that the `else` branch uses `ASTHelpers.getSymbol(leaf)` which i don't think will work for a `BlockTree`

File: nullaway/src/main/java/com/uber/nullaway/NullAway.java (hunk 4 above)
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/NullAway.java]: Yeah, the "right" thing to do here is to treat the field as `@Nullable` until it gets initialized.  That is a deeper change (to dataflow analysis, etc.), but probably the "right" way to do things.  For now I think we're ok with a narrow set of exclusions

File: nullaway/src/test/resources/com/uber/nullaway/testdata/ReadBeforeInitNegativeCases.java (hunk 3 above)
Comment (msridhar) [reply nullaway/src/test/resources/com/uber/nullaway/testdata/ReadBeforeInitNegativeCases.java]: Yeah I saw this in our code base; very cool local suppression technique!
